### DISCUSSION TITLE
catch all exceptions

### DISCUSSION THREAD

**Discussion by DJafari** [nutgram/nutgram#749] [Q&A] [answered]
is possible catch all exceptions to prevent bot webhook loop ?

i don't want use try catch for each request, just catch all exceptions in all of request of bot

**Comment by Lukasss93** [comment]
https://nutgram.dev/docs/usage/handlers#onexception

https://nutgram.dev/docs/usage/handlers#onapierror

**Reply by Perimental1** [reply]
Hmmm...

**Reply by DJafari** [reply]
i know it, but even when i define it, api requests throws again
i think normal behavior is when we defined exception handler for bot, request should not throws to exceptions

check this :

```php
$bot->onException(function (Nutgram $bot, \Throwable $exception) {
    var_dump('EXCEPTION');
});

$bot->sendMessage(
    text: 'Test',
    chat_id: 43235235235235235, // Wrong chat id
);
```

![image](https://github.com/user-attachments/assets/c5e3f952-9db9-47c7-8c6d-e5f69c40dbe4)

and so `var_dump('EXCEPTION'); ` never called

**Comment by Lukasss93** [comment, marked as answer]
The are 3 ways to handle exceptions:

1. try-catch (without using the onApiError handler)

```php
$bot = new Nutgram('token');

$bot->onCommand('start', function (Nutgram $bot) {
    try {
        $post = $bot->sendMessage('Hello', chat_id: 123456);
        echo gettype($post).PHP_EOL;
    } catch (Throwable $e){
        echo '[try] '.$e::class. ' => ' . $e->getMessage().PHP_EOL;
    }
});

$bot->run();
```

Result: 
>[try] SergiX44\Nutgram\Telegram\Exceptions\TelegramException => Forbidden: bot was blocked by the user

2. using the onAPiError handler

```php
$bot = new Nutgram('token');

$bot->onCommand('start', function (Nutgram $bot) {
    $post = $bot->sendMessage('Hello', chat_id: 123456);
    echo gettype($post).PHP_EOL;
});

$bot->onApiError(function (Nutgram $bot, TelegramException $e) {
    echo '[onApiError] '.$e::class. ' => ' . $e->getMessage().PHP_EOL;
});

$bot->run();
```

Result: 
>[onApiError] SergiX44\Nutgram\Telegram\Exceptions\TelegramException => Forbidden: bot was blocked by the user
>NULL

3. try-catch + registerApiException

```php
$bot = new Nutgram('token');

$bot->onCommand('start', function (Nutgram $bot) {
    try {
        $post = $bot->sendMessage('Hello', chat_id: 123456);
        echo gettype($post).PHP_EOL;
    } catch (Throwable $e){
        echo '[try] '.$e::class. ' => ' . $e->getMessage().PHP_EOL;
    }
});

$bot->registerApiException(BotBlocked::class);

$bot->run();
```

```php
class BotBlocked extends ApiException
{
    public static ?string $pattern = '.*bot was blocked.*';
}
```

Result:
>[try] App\BotBlocked => Forbidden: bot was blocked by the user

**Reply by DJafari** [reply]
```
$bot->onApiError(function (Nutgram $bot, TelegramException $e) {
    echo '[onApiError] '.$e::class. ' => ' . $e->getMessage().PHP_EOL;
});
```

this solved my question, thanks!
